internal/storage/storage.go

```diff
@@ -0,0 +1,220 @@
+
+	// ErrConflict indicates that the attempted write failed because of a version
+	// or UID mismatch.
+	ErrConflict = errors.New("operation failed because of a Version or Uid mismatch")
+
+	// ErrInconsistent indicates that the attempted write or consistent read could
+	// not be achieved because of a consistency or availability issue (e.g. loss of
+	// quorum, or when interacting with a Raft follower).
+	ErrInconsistent = errors.New("cannot satisfy consistency requirements")
+)
+
+// Backend provides the low-level storage substrate for resources. It can be
+// implemented using internal (i.e. Raft+MemDB) or external (e.g. DynamoDB)
+// storage systems.
+//
+// Refer to the method comments for details of the behaviors and invariants
+// provided, which are also verified by the conformance test suite in the
+// internal/storage/conformance package.
+//
+// Cross-cutting concerns:
+//
+// # UIDs
+//
+// Users identify resources with a name of their choice (e.g. service "billing")
+// but internally, we add our own identifier in the Uid field to disambiguate
+// references when resources are deleted and re-created with the same name.
+//
+// # GroupVersion
+//
+// In order to support automatic translation between schema versions, we only
+// store a single version of a resource, and treat types with the same Group
+// and Kind, but different GroupVersions, as equivalent.
+type Backend interface {
+	// Read a resource using its ID.
+	//
+	// # UIDs
+	//
+	// If id.Uid is empty, Read will ignore it and return whatever resource is
+	// stored with the given name. This is the desired behavior for user-initiated
+	// reads.
+	//
+	// If id.Uid is non-empty, Read will only return a resource if its Uid matches,
+	// otherwise it'll return ErrNotFound. This is the desired behaviour for reads
+	// initiated by controllers, which tend to operate on a specific lifetime of a
+	// resource.
+	//
+	// See Backend docs for more details.
+	//
+	// # GroupVersion
+	//
+	// If id.Type.GroupVersion doesn't match what is stored, Read will return a
+	// GroupVersionMismatchError which contains a pointer to the stored resource.
+	//
+	// See Backend docs for more details.
+	//
+	// # Consistency
+	//
+	// Read makes no guarantees about consistency, and may return stale results.
+	// For stronger guarantees, use ReadConsistent.
+	Read(ctx context.Context, id *pbresource.ID) (*pbresource.Resource, error)
+
+	// ReadConsistent provides the same functionality as Read, but guarantees
+	// single-resource sequential consistency, typically by bypassing any caches
+	// and proxying the request directly to the underlying storage system.
+	//
+	// If a consistent read cannot be achieved (e.g. when interacting with a Raft
+	// follower, or quorum is lost) ErrInconsistent will be returned.
+	//
+	// Use ReadConsistent sparingly, and prefer Read when possible.
+	ReadConsistent(ctx context.Context, id *pbresource.ID) (*pbresource.Resource, error)
```

Comment:
I like that a lot! I'm imagining some helpers methods like this:

```Go
resource, err := backend.Read(ctx, id, storage.MaxStaleness(30 * time.Second))

resource, err := backend.Read(ctx, id, storage.StronglyConsistent())
```

Couple of follow up questions:
1. Am I right that we don't actually have a way to measure the staleness of a Raft follower as a wall clock time? I guess the closest we could do would be to compare the applied index and make inferences from there.
2. Other than strong/eventual consistency and max staleness, can you think of any other parameters we'd want to support?
3. Do we want to maintain the difference between our "default" and "consistent" consistency modes? In https://github.com/hashicorp/consul/pull/16619 I've more-or-less abandoned the "default" mode, with the rationale being that we're going to assume results are stale and use expensive consistent reads sparingly (i.e. just for ACL enforcement).

Reply:
1. Some RPCs respect the `max_stale` query param today. They issue the query once with `AllowStale` being true, the response tacks on the time since the follower was last contacted by the leader. If the max stale duration exceeds that last contact time, the RPC will be retried against the leader instead: https://github.com/hashicorp/consul/blob/9d21736e9f5828920cb8302397e0f11c95cae88c/agent/catalog_endpoint.go#L250-L254

Comparing against the index would require knowing what index the leader was at. I _think_ we could glean that information from Raft to detect that its having a hard time catching up even if the leader is contacting the follower regularly.

2. I cannot think of anything else we would need right now. I also don't see adding more options as too difficult a task either so I would focus on what we know we need now.
3. In general I think the consistency modes will be up for interpretation by backends. When using Raft consistent means going to the leader whereas for leaderless backends consistent could mean retrieving from the underlying storage (k8s, dynamo etc) in a consistent manner. I need to review your other PR but my initial inclination would be to keep default mode as an option. Maybe there is a good reason why it doesn't work out cleanly and we need to drop it. I think your assumptions about our internal usage of the APIs being predominantly stale is correct. Whether that would hold for other users of the API I am less certain. I will come back to this after taking a look at your other PR.

Reply:
Thanks for the answers, Matt!

### Bounding Staleness
I remember we spoke about bounding staleness before in the context of the work @rboyer was doing to prevent old servers re-joining the cluster and handling RPCs. There are a couple of approaches we could take:

#### Stamp Raft log entries with the wall-clock time, and track the highest timestamp seen
Clock skew aside, this would be imprecise in clusters with light/infrequent write load (unless we pushed the timestamp out with the Raft heartbeat).

#### Estimate the number of writes per second
By estimating the leader is handling 100 writes/s, for example, a follower could infer that it is lagging `(<leader commit index> - <follower applied index>) / 100` seconds behind.

### Consistency Modes

I'd be in favor of supporting a smaller number of modes with very clearly defined guarantees that can be supported by all backends.

The Raft backend introduced in #16619 essentially supports two:
- **Eventually Consistent (default):** Equivalent to our current "stale" mode. Reads are [sequentially consistent](https://jepsen.io/consistency/models/sequential) with each other, but may not reflect writes made to the same server because of leader forwarding.
- **Strongly Consistent:** Equivalent to our current "consistent" mode. Full [linearizability](https://jepsen.io/consistency/models/linearizable).

The main difference between our current "default" and "consistent" modes is whether or not we do a [round of heartbeats](https://pkg.go.dev/github.com/hashicorp/raft#Raft.VerifyLeader) to verify leadership. In my opinion, the "default" mode only really makes sense if consistent reads are common enough to justify the reduction in safety for the sake of performance or scalability.

In the new controller architecture, we're assuming eventual consistency for the majority of our internal uses. Which I think makes it a reasonable assumption that if we opt into strong consistency, correctness really matters, so we should verify leadership.

You do raise a good point about external users of the API, though.

Reply:
Following up from our call earlier:

We'll need to support the old "default" mode for backward-compatibility with the v1 API. To that end, I think we'll have a single `Read` method with a consistency param like you suggested, and a "legacy" mode that's only supported by the Raft backend (and isn't available in the v2 API at all).

I think we can leave bounded staleness as a potential future improvement. As we discussed, it's difficult to implement something better than just checking the last leader contact timestamp (and probably even harder in external systems we have less control over).

Reply:
Sorry for all the back-and-forth on this topic 😅 

I realized shortly after writing my last comment, that if the v1 handlers:
- Continue to do their own leader forwarding
- Call the resource service in the same process (via something similar to our net/rpc in-mem codec)

We don't actually need to support a "legacy" mode at the storage backend level, because a leader handling an "eventually consistent" read provides the same guarantees as our current "default" mode 🎉

I also realized that the (intentional) omission of a strongly consistent version of `List` breaks backward compatibility. I don't think we want to support strongly consistent list operations going forward, because it requires serializability, but we probably need to support it in the Raft backend until the v1 APIs are gone.